src/HotChocolate/Core/src/Execution/RequestExecutor.cs

```diff
@@ -102,19 +146,37 @@ internal sealed class RequestExecutor : IRequestExecutor
         }
     }
 
-    public Task<IResponseStream> ExecuteBatchAsync(
-        IEnumerable<IQueryRequest> requestBatch,
-        bool allowParallelExecution = false,
-        CancellationToken cancellationToken = default)
+    private void DecrementActiveRequests(IExecutionResult? result)
     {
-        if (requestBatch is null)
+        if (result == null || !result.IsStreamResult())
         {
-            throw new ArgumentNullException(nameof(requestBatch));
+            Interlocked.Decrement(ref _activeRequests);
+            return;
         }
 
-        return Task.FromResult<IResponseStream>(
-            new ResponseStream(
-                () => _batchExecutor.ExecuteAsync(this, requestBatch),
-                ExecutionResultKind.BatchResult));
+        result.RegisterForCleanup(new DisposableAction(() =>
+        {
+            Interlocked.Decrement(ref _activeRequests);
+#if NETCOREAPP3_1 || NETSTANDARD2_0
+            return default;
+#else
+            return ValueTask.CompletedTask;
+#endif
+        }));
+    }
+
+    private class DisposableAction : IAsyncDisposable
```

Comment:
We can formulate this as a sync Action. `RegisterForCleanup` has four overloads.


```csharp
result.RegisterForCleanup(() =>
{
    Interlocked.Decrement(ref _activeRequests);
});
```

Or we formulate this as an object that implements IDisposable with finalizer where the object holds an instance if the request executor.

Reply:
@michaelstaib I'm not sure I can see the need for the finalizer and the IDisposable. Could you have a look at the updates there please? Definitely cleaner this way though good call!

Reply:
Sorry, just not sure if we need both there.

Reply:
Changes applied.